packages/vite/src/node/plugins/resolve.ts

```diff
@@ -142,6 +142,19 @@ export function resolvePlugin(resolveOptions: InternalResolveOptions): Plugin {
     name: 'vite:resolve',
 
     async resolveId(id, importer, resolveOpts) {
+      if (
+        id[0] === '\0' ||
+        id.startsWith('virtual:') ||
+        // When injected directly in html/client code
+        id.startsWith('/virtual:') ||
```

Comment:
I'm not sure you can inject this directly in html/client and work 🤔 Do you have an example of this?

Reply:
unoCSS for example: https://github.com/unocss/unocss/blob/main/packages/vite/src/devtool.ts#L141
(here it's not using the "/virtual:" prefix but it could)

Reply:
That means it could be anything though, and we shouldn't have a special preference for `/virtual:`?

Reply:
I think we should provide and encourage using a prefix that both:
- doesn't trigger fs calls if the plugin is not "pre"
- is a valid url and doesn't need to be pre-process by Vite

Without this line, the line you need to use `/@id/__x00__@unocss/devtools` which is not great compare to `/virtual:@unocss/devtools`

Reply:
The use of `\0` helps because the rollup ecosystem expects that when handling virtual modules (there are checks in rollup plugins when dealing with source maps for example). Uno should use `\0` in the resolved id here.

I think that converting `\0` to `/@id/__x00__` is making folks shy away from using the proper virtual marker. We could change our scheme in Vite 5 (or before?) to something that looks better. For example `/@virtual/` or `/@id/virtual:`. @antfu what do you think?

Reply:
It's always tricky to have conventions, as they are pollution other spaces and hard to convert them back. Since `virtual:` is valid in URL I guess we could ship it without changing to another format.

Reply:
We already have the `\0` convention though, so the `/@id/virtual:` would be just an internal implementation detail in Vite's server (replacing `/@id/__x00__`).

We can't use `virtual:` AFAICS without talking with rollup and rollup plugin authors to start adding checks for both `\0` and `virtual:` every time they check for a virtual file ([example in createFilter](https://github.com/rollup/plugins/blob/master/packages/pluginutils/src/createFilter.ts#L48), [example in Rollup's core](https://github.com/rollup/rollup/blob/c2cda424e69686671ba010d628c0f70c43a563f8/src/utils/options/normalizeInputOptions.ts#L138)).

I think `\0` is ok during build, and if we move from `/@id/__x00__` to something that looks nicer in the browser and logs during dev, it should be ok, no?

Reply:
I don't think it's worth the friction to move to something else. But I think going back to the initial discussion for `/virtual`, since this is rare that happens since html/client only, we could skip it for now? This `resolveId` function will run for many things and it would be good to avoid checking for things that don't happen often 🤔

Reply:
I agree to remove it until we agree on a pattern.
But when you look at profiling, import parsing & transformation starts to show up and I think it would be good to move to import patterns that are more ESM friendly

Reply:
What do you mean by ESM friendly? Having to encode \0 for the browser shouldnt have a noticeable perf hit, no?

Reply:
The perf hit is in parsing/rewriting every bit of JS we're sending to the browser.
And rewriting all import implies a lot of string manipulation where JS is not really fast.
I can't find a specific entry for that in a profile right now, but this is one of the main bottleneck on [rds](https://github.com/ArnaudBarre/rds).

But when you look at a profile, you see a lot of small things that are related to import transformation (new URL which is really slow in node, regex, ensureVersion, isOptimisedDep, ...)

If we rethink the patterns, we should go for things that does't need serialization and can directly go to as fs.read call when hitting back the server.

Reply:
I don't think we can avoid resolving the URLs though. We at least need to check them once. My point is that `\0` vs `/virtual:` doesn't affect this in particular. We still need to improve things while resolving, even after all we did in the past weeks, there seems to be a lot more to gain from making `vite:resolve` leaner. `new URL` is a good example, we should check if we can replace that by something faster.

About this PR, let's keep the new `/virtual:` condition, and remove the `'/@react-refresh' and we can merge it?